fl4health/parameter_exchange/sparse_coo_parameter_exchanger.py

```diff
@@ -0,0 +1,154 @@
+            score_gen_function (Callable[..., Dict[str, Tensor]]): Function that is responsible for
+            generating a score for every parameter inside a model in order to facilitate parameter selection.
+
+            In most cases, this function takes as inputs a current model and an initial model,
+            and it returns a dictionary that maps the name of each of the current model's tensors to
+            another tensor which contains the parameter scores.
+        """
+        assert 0 < sparsity_level <= 1
+        self.set_sparsity_level(sparsity_level)
+        self.parameter_packer: SparseCooParameterPacker = SparseCooParameterPacker()
+        self.set_score_gen_function(score_gen_function)
+
+    def set_sparsity_level(self, sparsity_level: float) -> None:
+        """This method exists to allow for setting new sparsity levels throughout training."""
+        self.sparsity_level = sparsity_level
+
+    def set_score_gen_function(self, score_gen_function: Callable[..., Dict[str, Tensor]]) -> None:
+        self.score_gen_function = score_gen_function
+
+    def generate_parameter_scores(self, model: nn.Module, initial_model: Optional[nn.Module]) -> Dict[str, Tensor]:
+        """Calling the score generating function to produce parameter scores."""
+        return self.score_gen_function(model, initial_model)
+
+    def create_masks(self, model: nn.Module, initial_model: Optional[nn.Module]) -> Dict[str, Tensor]:
+        """
+        Produce masks for model's tensors.
+
+        More precisely, this method first leverage a score generating function
+        to generate scores for all parameters of model.
+
+        Next, these scores are used to produce masks.
+        A threshold is determined according to the desired sparsity level,
+        then parameters whose scores are less than this threshold are assigned a mask value of 0,
+        while parameters whose scores are greater than or equal to this threshold are
+        assigned a mask value of 1.
+
+        Args:
+            model (nn.Module): Current model.
+            initial_model (nn.Module): Initial model.
+
+        Returns:
+            Dict[str, Tensor]: Dictionary that maps the name of each tensor to its corresponding mask.
+        """
+        all_parameter_scores = self.generate_parameter_scores(model, initial_model)
+        all_scores = torch.cat([val.flatten() for _, val in all_parameter_scores.items()])
+        # Sorting all scores and determining the threshold.
+        sorted_scores, _ = torch.sort(all_scores, descending=True)
+        top_scores = sorted_scores[: math.ceil(len(sorted_scores) * self.sparsity_level)]
+        score_threshold = top_scores[-1].item()
+
+        names_to_masks = {}
+        for name, param_scores in all_parameter_scores.items():
+            # Use score_threshold to produce masks.
+            threshold_result = torch.where(param_scores >= score_threshold, input=param_scores, other=0)
```

Comment:
The mask is created by thresholding the parameter scores. So in theory, if all parameters have the same score (for example, when the model parameters are one constant) then the number of parameters selected won't necessarily match the sparsity level. But I think it is reasonable to assume that this won't happen in practice.

Reply:
Is it worth checking if all elements are equal with torch.unique or something and logging a warning or raising an error if this is the case? Im thinking it would be helpful to surface if the provided scoring function produces the same value cause it might caused by an underlying bug